src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -836,8 +836,9 @@ public void recover(RecoveryHandler recoveryHandler) throws EngineException {
             maybeFailEngine("recovery phase 3", e);
             throw new RecoveryEngineException(shardId, 3, "Execution failed", wrapIfClosed(e));
         } finally {
-            Releasables.close(success, onGoingRecoveries, writeLock, phase1Snapshot,
-                    phase2Snapshot, phase3Snapshot); // hmm why can't we use try-with here?
+            // close the snapshots first to release the reference to the translog file, so a flush post recovery can delete it
+            Releasables.close(success, phase1Snapshot, phase2Snapshot, phase3Snapshot,
+                    onGoingRecoveries, writeLock); // hmm why can't we use try-with here?
```

Comment:
it in-fact might make sense to have the write lock in an outer try catch to ensure it's released last?